posts/2023-12-21-async-fn-rpit-in-traits.md

```diff
@@ -0,0 +1,229 @@
+}
+```
+
+## Where the gaps lie
+
+### `-> impl Trait` in public traits
+
+The use of `-> impl Trait` is still discouraged for general use in **public** traits, for the reason that users can't use additional bounds on the return type. For example, there is no way to write this function:
+
+```rust!
+fn print_in_reverse(container: impl Container) {
+    for item in container.items().rev() {
+        // ERROR:                 ^^^
+        // the trait `DoubleEndedIterator`
+        // is not implemented for
+        // `impl Iterator<Item = Widget>`
+        eprintln!("{item}");
+    }
+}
+```
+
+In the future we plan to add a solution for this. For now, this feature is best used in internal traits or when you're confident your users won't need additional bounds like this. Otherwise, you should continue using associated types.
+
+### `async fn` in public traits
+
+Since `async fn` desugars to `-> impl Future`, the same limitations apply. If you use bare `async fn` in a public trait today, you'll see a warning. Thankfully, we have a solution that allows using `async fn` in public traits today!
+
+```
+warning: use of `async fn` in public traits is discouraged as auto trait bounds cannot be specified
+ --> src/lib.rs:7:5
+  |
+7 |     async fn fetch(&self, url: Url) -> HtmlBody;
+  |     ^^^^^
+  |
+help: you can desugar to a normal `fn` that returns `impl Future` and add any desired bounds such as `Send`, but these cannot be relaxed without a breaking API change
+  |
+7 -     async fn fetch(&self, url: Url) -> HtmlBody;
+7 +     fn fetch(&self, url: Url) -> impl std::future::Future<Output = HtmlBody> + Send;
+  |
+```
+
+Of particular interest to users of async are `Send` bounds on the returned future. Since users cannot add bounds later, the error message is saying that you as a trait author need to make a choice: Do you want your trait to work with multithreaded, work-stealing executors?
+
+For public traits, we recommend using the `trait-variant` proc macro to let your users choose. Add it to your project with `cargo add trait-variant`, then use it like so:
+
+```rust
+#[trait_variant::make(HttpService: Send)]
+pub trait LocalHttpService {
+    async fn fetch(&self, url: Url) -> HtmlBody;
+}
+```
+
+This creates *two* versions of your trait: `LocalHttpService` for single-threaded executors and `HttpService` for multithreaded work-stealing executors. Since we expect the latter to be used more commonly, it has the shorter name in this example. It has additional Send bounds:
+
+```rust
+pub trait HttpService: Send {
+    fn fetch(
+        &self,
+        url: Url,
+    ) -> impl Future<Output = HtmlBody> + Send;
+}
+```
+
+This macro works for async because `impl Future` rarely requires additional bounds other than Send, so we can set our users up for success. See [this blog post](https://blog.rust-lang.org/inside-rust/2023/05/03/stabilizing-async-fn-in-trait.html) for a more thorough explanation of the problem.[^cut-scope]
```

Comment:
```suggestion
This macro works for async because `impl Future` rarely requires additional bounds other than `Send`, so we can set our users up for success. See [this blog post](https://blog.rust-lang.org/inside-rust/2023/05/03/stabilizing-async-fn-in-trait.html) for a more thorough explanation of the problem.[^cut-scope]
```